### DISCUSSION TITLE
missing module (logging.config)

### DISCUSSION THREAD

**Discussion by Shnitzelil** [mitmproxy/mitmproxy#6624] [Q&A] [answered]
Hi,
I'm trying to use spnego 3rd party to establish ntlm authentication (the ntlm_auth is deprecated).
Anyway, when it trying to initialize it fails saying "ModuleNotFoundError: No module named 'logging.config'."
How can I make it work?
Thank you

**Comment by Prinzhorn** [comment]
Please provide steps to reproduce the error so we know what you are trying to do

**Reply by Shnitzelil** [reply]
In general,
Implementing addon that have import spengo and configuring mitmdump to load the new addon.
I based my solution on the ntlm_upstream_proxy.py from contrib folder. https://github.com/mitmproxy/mitmproxy/blob/main/examples/contrib/ntlm_upstream_proxy.py

**Reply by Prinzhorn** [reply]
Does `ntlm_upstream_proxy.py` produce the same error? Or only your own addon?

**Reply by Shnitzelil** [reply]
no. 
Sadly the ntlm_upstream_proxy use ntlm_auth (a deprecated 3rd party).
So, I replaced it with spengo (suggested in the ntlm_auth Readme https://github.com/jborean93/ntlm-auth).
Anyway, my question is how can I add the missing module (present in Python SDK).

**Reply by Prinzhorn** [reply, marked as answer]
The error message I'm getting with the following addon:

```py
import logging.config
```

Pretty much gives you the answer, right?

```
ModuleNotFoundError: No module named 'logging.config'.

Note that mitmproxy's binaries include their own Python environment. If your addon requires the installation of additional dependencies, please install mitmproxy from PyPI (https://docs.mitmproxy.org/stable/overview-installation/#installation-from-the-python-package-index-pypi).
```

**Reply by Shnitzelil** [reply]
Thank you for checking it.

The link explains how to add 3rd party to the mitmproxy and (AFAIK) the logging.config is part of Python SDK.
Please help me to understand what I'm missing here.

Thanks

**Reply by Prinzhorn** [reply]
> Note that mitmproxy's binaries include **their own Python environment**

Which apparently doesn't include `logging.config`, because it's not used by mitmproxy. I'm the wrong person to ask how exactly that works, I assume some PyInstaller magic that reduces the binary size by only including things that are actually used.

Edit: it might be worth opening an issue for this, because if I understand correctly you have no way of installing it, since it should be built in. Maybe we need to ship mitmproxy with logging.config (again, I have no idea how any of this works)

**Comment by Prinzhorn** [comment]
You might get away with this:

```py
import sys

sys.modules["logging.config"] = {}
```

Before any other imports. Because you might not need it:

https://github.com/jborean93/pyspnego/blob/cba319d9795148702ce6d5575f13508a12fab4ec/src/spnego/__init__.py#L50-L57

**Reply by Shnitzelil** [reply]
Will check that.
Thanks!

**Reply by Shnitzelil** [reply]
Ok, looks like it works.
I saw that you wrote that you are not the right person.
Yet, I wonder what it will required to add this module to be in mitm binary?

Thanks

**Reply by mhils** [reply]
We're not going to add modules that are not used by mitmproxy itself because that leads to binary bloat. If you need a full Python environment, the blessed way is to install from PyPI.
